bundles/org.openhab.binding.caddx/src/main/java/org/openhab/binding/caddx/internal/CaddxCommunicator.java

```diff
@@ -0,0 +1,475 @@
+            {
+                loopUntilByteIsRead(in, 0x7e);
+
+                inMessage = true;
+                messageBufferLength = 0;
+            }
+            logger.trace("CaddxCommunicator.handleBinaryProtocol() Got start byte");
+
+            // Read the message length
+            if (messageBufferLength == 0) {
+                int b = readByte(in);
+                messageBufferLength = b + 2; // add two bytes for the checksum
+                message = new byte[messageBufferLength];
+                messageBufferIndex = 0;
+            }
+            logger.trace("CaddxCommunicator.handleBinaryProtocol() Got message length {}", messageBufferLength);
+
+            // Read the message
+            do {
+                int b = readByte(in);
+                message[messageBufferIndex] = (byte) b;
+
+                if (message[messageBufferIndex] == 0x7D) {
+                    unStuff = true;
+                    continue;
+                }
+
+                if (unStuff) {
+                    message[messageBufferIndex] |= 0x20;
+                    unStuff = false;
+                }
+
+                messageBufferIndex++;
+            } while (messageBufferIndex < messageBufferLength);
+
+            // Offer the message
+            offerCaddxMessage();
+
+            logger.trace("CaddxCommunicator.handleBinaryProtocol() Got message {}", message[0]);
+        } catch (EOFException e) {
+            return;
+        } catch (IOException e) {
+        } catch (InterruptedException e) {
+            logger.trace("InterruptedException caught.");
+            Thread.currentThread().interrupt();
+        }
+
+        // Initialize state for a new reception
+        inMessage = false;
+        messageBufferLength = 0;
+        messageBufferIndex = 0;
+        unStuff = false;
+    }
+
+    private void receiveInAsciiProtocol(SerialPortEvent serialPortEvent) {
+        try {
+            // Read the start byte
+            if (!inMessage) {
+                loopUntilByteIsRead(in, 0x0a);
+
+                inMessage = true;
+                haveFirstByte = false;
+                messageBufferLength = 0;
+            }
+            logger.trace("CaddxCommunicator.handleAsciiProtocol() Got start byte");
+
+            // Read the message length
+            if (messageBufferLength == 0) {
+                int b = readAsciiByte(in);
+                messageBufferLength = b + 2; // add 2 bytes for the checksum
+                message = new byte[messageBufferLength];
+            }
+            logger.trace("CaddxCommunicator.handleAsciiProtocol() Got message length {}", messageBufferLength);
+
+            // Read the message
+            do {
+                int b = readAsciiByte(in);
+                message[messageBufferIndex] = (byte) b;
+                messageBufferIndex++;
+            } while (messageBufferIndex < messageBufferLength);
+
+            // Offer the message
+            offerCaddxMessage();
```

Comment:
Is it safe to block the SerialPortEvent dispatcher thread for 3 seconds while trying to offer the message to the synchronous queue? What if another SerialPortEvent comes by during that time? Does it execute on another thread?

Reply:
Communication protocol:
One party (panel or binding) sends a message and the other party has to acknowledge within 3 seconds. If the acknowledgement does not arrive the initial party re-sends the message.

When a full message is retrieved by the binding, it has 3 seconds to send the reply. During this period the panel does not send another message.
The SerialPortEvent that will potentially arrive in these 3 seconds will not be DATA_AVAILABLE (and is ignored). I think it is safe (and expected) to block for 3 seconds.

Reply:
Resolving this